src/sage/rings/finite_rings/integer_mod_ring.py

```diff
@@ -1547,6 +1551,414 @@ def random_element(self, bound=None):
+            []
+
+            sage: R = Zmod(11**3)
+            sage: S.<x> = R[]
+            sage: f = x^2 + 123*x + 1
+            sage: R._lift_residue_field_root(11, 3, f, f.derivative(), Zmod(11)(10))
+            [10, 131, 252, 373, 494, 615, 736, 857, 978, 1099, 1220, 109, 230, 351, 472, 593, 714, 835, 956, 1077, 1198, 1319]
+        """
+        if e == 1:
+            # Nothing to do
+            return (root,)
+        deriv = fprime(root)
+        if deriv:
+            # Unique lift, use Newton iteration
+            prec = 1
+            while True:
+                prec = min(2*prec, e)
+                Zp_prec = Zmod(p**prec)
+                root = Zp_prec(root.lift())
+                deriv = fprime(root)
+                step = f(root) / deriv
+                root -= step
+                if prec >= e:
+                    return (root,)
+        else:
+            # Non-unique lift, go one power at a time
+            prec = 1
+            new_power = 1
+            new_mod = p
+            current_roots = (root,)
+            for _ in range(e - 1):
+                prec += 1
+                new_power = new_mod
+                new_mod *= p
+                new_roots = []
+                Zp_prec = Zmod(new_mod)
+                for rt in current_roots:
+                    rt = Zp_prec(rt.lift())
+                    if f(rt):
+                        continue
+                    new_roots.append(rt)
+                    for _ in range(p - 1):
+                        rt += new_power
+                        new_roots.append(rt)
+                current_roots = new_roots
+
+            return current_roots
+
+    def _roots_univariate_polynomial(self, f, ring=None, multiplicities=True, algorithm=None):
+        r"""
+        Return the roots of ``f`` in the ring ``ring``.
+
+        INPUT:
+
+        - ``f`` - a polynomial defined over this ring
+
+        - ``ring`` - the ring to find roots in.  Control flow elsewhere
+          ensures that the only cases we need to handle are ``self`` and
+          ``None``.  Otherwise we raise ``NotImplementedError``.
+
+        - ``multiplicities`` - bool (default: ``True``). If ``True``, return
+          list of pairs `(r, n)`, where `r` is a root and `n` is its
+          multiplicity. If ``False``, just return the unique roots, with no
+          information about multiplicities. Multiplicities are only defined
+          over fields, and this method raises ``NotImplementedError`` if this
+          is ``True`` but the ring is not a field.
+
+        - ``algorithm`` - ignored
+
+        ALGORITHM:
+
+        The algorithm is adapted from [Gouvea], section 4.5, and [Coh1993],
+        section 3.5.3.  It is a combination of the Chinese Remainder Theorem
+        and Hensel's lemma.  As a base case, if $N$ is prime, then we find
+        roots by factoring $f$.  If $N$ is a prime power $p^e$, then we find
```

Comment:
See the [developer guide](https://doc.sagemath.org/html/en/developer/coding_basics.html#section-latex-typeset) for how to do LaTeX formatting in our docstrings and change the formatting accordingly.